conda/__init__.py

```diff
@@ -148,3 +139,209 @@ def _default(self, obj):
 
 _default.default = JSONEncoder().default
 JSONEncoder.default = _default
+
+
+# inspired by deprecation (https://deprecation.readthedocs.io/en/latest/) and
+# CPython's warnings._deprecated
+class _deprecated:
+    _version: tuple[int, ...]
+    _category: Warning
+    _message: Callable[str, str]
+    _argument: str | None = None
+    _rename: str | None = None
+    _stack: int
+
+    @staticmethod
+    def _parse_version(version: str) -> tuple[int, ...]:
+        """A naive version parser to avoid circular imports. Do not use for other purposes."""
+
+        def try_int(string: str) -> int | str:
+            try:
+                return int(string)
+            except ValueError:
+                return string
+
+        return tuple(map(try_int, version.split(".")))
+
+    def __init__(
+        self,
+        deprecate_in: str,
+        remove_in: str,
+        *,
+        addendum: str | None = None,
+        stack: int = 0,
+    ):
+        """Deprecation decorator for functions, methods, & classes.
+
+        Args:
+            deprecate_in: Version in which code will be marked as deprecated.
+            remove_in: Version in which code is expected to be removed.
+            addendum: Optional additional messaging. Useful to indicate what to do instead.
+            stack: Optional stacklevel increment.
+        """
+        deprecate_version = self._parse_version(deprecate_in)
+        remove_version = self._parse_version(remove_in)
+
+        addendum = f" {addendum}" if addendum else ""
+        if self._version < deprecate_version:
+            self._category = PendingDeprecationWarning
+            message = f"{{name}} is pending deprecation and will be removed in {remove_in}."
+        elif self._version < remove_version:
+            self._category = DeprecationWarning
+            message = f"{{name}} is deprecated and will be removed in {remove_in}."
+        else:
+            self._category = None
+            message = f"{{name}} was slated for removal in {remove_in}."
+        self._message = lambda name: f"{message}{addendum}".format(name=name)
```

Comment:
I'm a little worried by the way this is written, the mix of f-strings and old-style `str.format()`ing etc. Can we just use this?

```suggestion

        if self._version < deprecate_version:
            category = PendingDeprecationWarning
            warning = f"is pending deprecation and will be removed in {remove_in}."
        elif self._version < remove_version:
            category = DeprecationWarning
            warning = f"is deprecated and will be removed in {remove_in}."
        else:
            category = None
            warning = f"was slated for removal in {remove_in}."

        def message(name):
            parts = [name, warning]
            if addendum:
                parts.append(addendum)
            return " ".join(parts)

        self._category = category
        self._message = lambda name: message(name)

```

Reply:
This too was an area I wasn't too happy with, these are some good ideas, what about this:

```suggestion
        if self._version < deprecate_version:
            category = PendingDeprecationWarning
            warning = f"is pending deprecation and will be removed in {remove_in}."
        elif self._version < remove_version:
            category = DeprecationWarning
            warning = f"is deprecated and will be removed in {remove_in}."
        else:
            category = None
            warning = f"was slated for removal in {remove_in}."

        self._category = category
        self._message = lambda name: f"{name} {warning} {addendum}".strip()
```

Reply:
Works for me!